### PR TITLE
Add EMS provisioning/deprovisioning

### PR DESCRIPTION
**Pull request by radufa**
**Description**

Extend KEB to provision/deprovision EMS

Changes proposed in this pull request:

- add for EMS new steps for: offering, provisioning, binding, unbinding, deprovisioning
- all EMS steps are disabled for now
- extend KEB model for EMS data
- persist EMS overrides as encrypted strings
- extend service manger "fake" implementation used by EMS unit tests
- add "integration" tests for EMS provisioning/deprovisioning which use a real environment (no mocks)

**Related issue(s)**

See: https://github.com/kyma-project/kyma/issues/9773

### CODE REVIEW

File: components/kyma-environment-broker/internal/process/provisioning/ems_bind.go [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+var _ Step = (*EmsBindStep)(nil)
+
+func (s *EmsBindStep) Name() string {
+	return "EMS_Bind"
+}
+
+func (s *EmsBindStep) Run(operation internal.ProvisioningOperation, log logrus.FieldLogger) (internal.ProvisioningOperation, time.Duration, error) {
+	if operation.Ems.Instance.InstanceID == "" {
+		log.Warnf("Ems Provisioning step was not triggered")
+		return operation, 0, nil
```
Comment (piotrmiskiewicz) [inline components/kyma-environment-broker/internal/process/provisioning/ems_bind.go:44]: such return statement means everything ok, step finishes. Is it ok?

File: components/kyma-environment-broker/internal/process/provisioning/ems_provision.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+	if operation.Ems.Instance.InstanceID != "" {
+		return operation, 0, nil
+	}
+
+	smCli, err := operation.ServiceManagerClient(log)
+	if err != nil {
+		return s.handleError(operation, err, log, "unable to create Service Manage client")
+	}
+	// get offering
+	offs, err := smCli.ListOfferingsByName(EmsOfferingName)
```
Comment (piotrmiskiewicz) [inline components/kyma-environment-broker/internal/process/provisioning/ems_provision.go:46]: there is a step for taking the offering IDs - ServiceManagerOfferingStep

File: components/kyma-environment-broker/internal/process/provisioning/ems_bind.go (hunk 1 above)
Comment (piotrmiskiewicz) [reply components/kyma-environment-broker/internal/process/provisioning/ems_bind.go:44]: you need to chech the last operation - if the provisioning has beed finished

Comment (radufa) [reply components/kyma-environment-broker/internal/process/provisioning/ems_bind.go]: Changed

File: components/kyma-environment-broker/internal/process/provisioning/ems_provision.go (hunk 2 above)
Comment (radufa) [reply components/kyma-environment-broker/internal/process/provisioning/ems_provision.go]: Refactored

File: components/kyma-environment-broker/internal/process/provisioning/ems_bind.go [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+	smCli, err := operation.ServiceManagerClient(log)
+	if err != nil {
+		return s.handleError(operation, err, log, fmt.Sprintf("Step %s : unable to create Service Manage client", s.Name()))
+	}
+	// test if the provisioning is finished, if not, retry after 10s
+	resp, err := smCli.LastInstanceOperation(operation.Ems.Instance.InstanceKey(), "")
+	if err != nil {
+		return s.handleError(operation, err, log, fmt.Sprintf("Step %s : LastInstanceOperation() call failed", s.Name()))
+	}
+	log.Infof("Step %s : Provisioning Ems (instanceID=%s) state: %s", s.Name(), operation.Ems.Instance.InstanceID, resp.State)
```
Comment (piotrmiskiewicz) [inline components/kyma-environment-broker/internal/process/provisioning/ems_bind.go:61]: the `log` instance contains data like "step name" so you don't need to put it in the message.

File: components/kyma-environment-broker/internal/runtime/components/components.go [hunk 4]
```diff
@@ -11,4 +11,5 @@ const (
 	Backup                  = "backup"
 	BackupInit              = "backup-init"
 	AvSBridge               = "avs-bridge"
+	Eventing                = "eventing"
```
Comment (piotrmiskiewicz) [inline components/kyma-environment-broker/internal/runtime/components/components.go:14]: where this const is used?

Comment (radufa) [reply components/kyma-environment-broker/internal/runtime/components/components.go:14]: It's used in "components/kyma-environment-broker/internal/process/provisioning/ems_bind.go"
see the "append overrides" call: https://github.com/kyma-project/control-plane/pull/386/files#diff-5199cf5d72391ebe9b2cc6594b136952d74a1ea72ba8e1d2709ccbad28e2bc4cR100

File: components/kyma-environment-broker/internal/process/provisioning/ems_bind.go (hunk 3 above)
Comment (radufa) [reply components/kyma-environment-broker/internal/process/provisioning/ems_bind.go:61]: The code on line 61 generates the following log message:
```
  time="2020-12-21T17:20:29+01:00" level=info msg="Step EMS_Bind : Provisioning Ems (instanceID=ff9a4806-abca-41ea-94e3-2ac7a139e38a) state: in progress"
```
Without having "Step %s:..." at the beginning of the message and "s.Name()" as a variable, no step name was logged.
This log extract was taken from running of integration test.

I've removed the "step name" from the message, cause I saw that the "manager" will set it before executing the step on production:
```
		for _, step := range steps {
			logStep := logOperation.WithField("step", step.Name())
			logStep.Infof("Start step")

			processedOperation, when, err = m.runStep(step, processedOperation, logStep)
```
 The superfluous "step name" was removed from all log messages.

Comment (radufa) [reply components/kyma-environment-broker/internal/process/provisioning/ems_bind.go:61]: Done

**Comment by piotrmiskiewicz** [comment]
one thing left - the initialisation.go (deprovisioning step) contains the following line:
```
operation.XSUAA = op.XSUAA
```
It takes XSUAA data from provisioning operation and puts it into the deprovisioning one. The same must be done for ems.

**Comment by radufa** [comment]
Done, now the Ems data from provisioning operation are also put into the deprovisioning operation.

**Review by piotrmiskiewicz — APPROVED** [review]

**Pull request merged**